### DISCUSSION TITLE
Backdrop filter not working after build

### DISCUSSION THREAD

**Discussion by designchapnl** [tailwindlabs/tailwindcss#7044] [Help]
**What version of Tailwind CSS are you using?**

v3.0.13

**What build tool (or framework if it abstracts the build tool) are you using?**

Nuxt.js v2.14.12

**What version of Node.js are you using?**

v16.13.0

**What browser are you using?**

Chrome, Safari

**What operating system are you using?**

MacOs Monterey v12.1

**Reproduction URL**

-

**Describe your issue**

I have a div on top of an image with backdrop blur applied to it. During development everything looks fine. After building, the backdrop filter blur isn't working, but the css is there. Not tested with other filters.

I'm using the class backdrop-blur-lg and generates:

`.backdrop-blur-lg {
    --tw-backdrop-blur: blur(16px);
    -webkit-backdrop-filter: var(--tw-backdrop-blur) var(--tw-backdrop-brightness) var(--tw-backdrop-contrast) var(--tw-backdrop-grayscale) var(--tw-backdrop-hue-rotate) var(--tw-backdrop-invert) var(--tw-backdrop-opacity) var(--tw-backdrop-saturate) var(--tw-backdrop-sepia);
    backdrop-filter: var(--tw-backdrop-blur) var(--tw-backdrop-brightness) var(--tw-backdrop-contrast) var(--tw-backdrop-grayscale) var(--tw-backdrop-hue-rotate) var(--tw-backdrop-invert) var(--tw-backdrop-opacity) var(--tw-backdrop-saturate) var(--tw-backdrop-sepia);
}`

At the end when I delete the undefined filters from the css in the console (I only use backdrop blur), the backdrop filter does its job again (Chrome and Safari).

**Comment by simonswiss** [comment]
Hey!

Do you have a repo where we can have a look? Is the `@tailwind base;` directive present in your CSS file? It's responsible to set some CSS variables, including stuff for filters, so make sure it's there.

**Reply by Lordnibbler** [reply]
I see a similar bug with `invert` and `brightness` filters at Tailwind 3.0.12 and above, also only building in production (development looks fine). I am using webpack/webpacker with Rails 7.0.1. 

The `invert` and `brightness` filters work properly in Tailwind 3.0.11. 

I suspect one of the changes in 3.0.12 broke some of these filters including the `backdrop-blur` mentioned by @designchapnl, as I see the production CSS output has changed without me making any change to my tailwind config, my CSS, or HTML markup.

3.0.12:
<img width="204" alt="image" src="https://user-images.githubusercontent.com/199422/149873994-fd762428-ec0b-4f03-9384-02137c3fa2b3.png">

3.0.11:
<img width="210" alt="image" src="https://user-images.githubusercontent.com/199422/149873998-24f201ea-173b-4e46-b64e-e61d89beb536.png">

**Reply by designchapnl** [reply]
Yes, exactly this. Thank you!

**Comment by designchapnl** [comment]
Hi!

These are the first lines of the css:

@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {

**Reply by little3201** [reply]
invert not work on safari after build

**Comment by willcosgrove** [comment]
I think I've figured out the source of the problem. At least it was the source of the problem for me 😇 

In production, my tailwind code was getting run through sass, and sass was tossing out the last backdrop filter variable assignment for some unknown reason.

So I should have had:

```css
*, ::before, ::after {
  --tw-backdrop-blur: ;
  --tw-backdrop-brightness: ;
  --tw-backdrop-contrast: ;
  --tw-backdrop-grayscale: ;
  --tw-backdrop-hue-rotate: ;
  --tw-backdrop-invert: ;
  --tw-backdrop-opacity: ;
  --tw-backdrop-saturate: ;
  --tw-backdrop-sepia ;
}
```

But in my case it left out the sepia variable. So when the browser was trying to resolve `var(--tw-backdrop-sepia)` inside the `backdrop-filter` declaration, it was essentially erroring, and setting the property to `none`.

I could have fixed it a couple of different ways. I could have manually declared the --tw-backdrop-sepia variable. But what I ended up doing was removing sass from the pipeline and going with an all postcss pipeline which didn't have the bug. Now that all the backdrop variables are initialized, the class is working as intended.

**Reply by Lordnibbler** [reply]
What happens if you downgrade to 3.0.11, but keep sass?

**Reply by willcosgrove** [reply]
3.0.11 works, but it's because 3.0.12 added the variable declarations that sass was mangling.

This is the PR that introduced the changes that are causing the problems listed here: #6926.

But I don't think the problem we're all experiencing is due to a bug in that PR, it's due to a bug elsewhere in our respective build pipelines. My reasoning for this is that if you strip away everything else, and just use the Tailwind CDN, it works fine:  https://play.tailwindcss.com/My70K5zys8

**Reply by Lordnibbler** [reply]
I see, good information here. I'll play around a bit, but unfortunately for me, I cannot remove sass as I am using tailwind in conjunction with a rails app. rails uses webpacker, which in its current version [lists `sass` as a dependency](https://github.com/rails/webpacker/blob/e0c998e2aa7f096709eaa2a7f2d7e29d413abed3/package.json#L45)

looks like webpacker v6, which is due out soon, drops sass as a dependency, although I'll still have issues as some of the libraries I bring in to my app contain sass files.

**Reply by willcosgrove** [reply]
I'm using webpacker as well. You don't have to remove sass from your dependencies, you just need to remove it from the build pipeline that handles tailwind. So the sass loader only handles `*.sass` and `*.scss` files, so I have my tailwind file as just a `.css` file. In that case it's being handled by a postcss loader.

I'll see if I can get you more details on my exact setup tomorrow when I'm back in front of it.

**Reply by willcosgrove** [reply]
@Lordnibbler so I'm using webpacker `6.0.0.beta.7`. I don't know if that has any bearing on the particular config I have. But it looks like on that version, it's configured with a sass loader, but only if you use a `.scss` or `.sass` extension on your file. If you just use `.css` then it only uses a postcss loader.

Then I have a tailwind.js file, which is an entrypoint file (or pack file, I forget the webpacker term used). The tailwind.js just imports a tailwind.css file, and the tailwind.css file has the `@import`s in them to load tailwind.

**Reply by Lordnibbler** [reply]
Oh, I understand. Yes, that is how I have tailwind build configured as well, just a simple `.css` file, not using sass there.

**Comment by winexy** [comment]
Had the same issue. In production build some of the CSS custom properties were removed

![image](https://user-images.githubusercontent.com/26577190/151040567-d8dcb526-395b-4940-81e5-79e9e483c386.png)

Then i've noticed that during production build **Vite** is logging some warnings 

![image](https://user-images.githubusercontent.com/26577190/151040424-c21b3572-49ca-4426-b2d1-932800bad29e.png)

Updated **Vite** from `v2.1.5` to `v2.7.13` (latest at the moment). After that problem disappeared

**Comment by thecrypticace** [comment]
Hi all! This appears to be, at least in some of the cases, a bug in postcss parsing that doesn't mix well with some cssnano optimizations. Can you upgrade your projects to the latest versions of postcss (>= 8.4.6) and cssnano and see if the problem disappears?

**Reply by designchapnl** [reply]
Unfortunately no change in build result after updating postcss.

**Reply by thecrypticace** [reply]
ah that's a bummer so it appears that more than just an postcss + cssnano issue. As far as Tailwind CSS goes it is generating the appropriate CSS in dev or production builds. So other PostCSS plugins, build tooling, or minification appears to be the issue here.

**Comment by liydaco** [comment]
Downgrading to 3.0.10 worked for me.

**Reply by daguitosama** [reply]
Downgrading to 3.0.10 worked out nicely, `blurs` back to production build.

**Comment by felquis** [comment]
In my use case I'm not using tailwind base because it saves me some bytes,

### Example

Remember to use a background with an alpha, like `rgba(255,255,255, 0.8)`

I'm only using this:
```html
<div className="backdrop-blur-md backdrop-saturate-50 bg-[rgba(255,255,255,0.8)]">
```

tailwind 3.0.21 adds this snippet:

```css
.backdrop-saturate-50 {
    --tw-backdrop-saturate: saturate(.5);
}
.backdrop-blur-md, .backdrop-saturate-50 {
    backdrop-filter: var(--tw-backdrop-blur) var(--tw-backdrop-brightness) var(--tw-backdrop-contrast) var(--tw-backdrop-grayscale) var(--tw-backdrop-hue-rotate) var(--tw-backdrop-invert) var(--tw-backdrop-opacity) var(--tw-backdrop-saturate) var(--tw-backdrop-sepia);
}
.backdrop-blur-md {
    --tw-backdrop-blur: blur(12px);
}
```

On this print bellow, devtools shows in blue color the custom properties applied to the backdrop-filter, just two out of nine.

![Screen Shot 2022-02-14 at 07 58 20](https://user-images.githubusercontent.com/736728/153851495-9a712867-20ad-4ecc-9adb-c7c4a7ecf1ee.png)

1. tailwindcss JIT should know I'm only using two backdrop-filters and strip-out all others properties for the sake of saving bytes.
2. The print shows only two properties are applied, tailwind does define the two custom properties I use, but it correctly does not add the others, so it breaks because the `backdrop-filter` definitions has a kind of indirect dependency on base styles.

The snippet above does not work as expected, but this script bellow does work:
```css
.backdrop-saturate-50 {
    --tw-backdrop-saturate: saturate(.5);
}
.backdrop-blur-md, .backdrop-saturate-50 {
    backdrop-filter: var(--tw-backdrop-blur) var(--tw-backdrop-saturate);
}
.backdrop-blur-md {
    --tw-backdrop-blur: blur(12px);
}
```

----

### Work-around-ish

Pre-defined all backdrop related custom properties even though your project doesn't need them, so the backdrop-filter from tailwindcss will evaluate the correct CSS expression.

```
*,
::before,
::after {
  --tw-backdrop-blur: ;
  --tw-backdrop-brightness: ;
  --tw-backdrop-contrast: ;
  --tw-backdrop-grayscale: ;
  --tw-backdrop-hue-rotate: ;
  --tw-backdrop-invert: ;
  --tw-backdrop-opacity: ;
  --tw-backdrop-saturate: ;
  --tw-backdrop-sepia: ;
}
```

Problem "solved", what do you think?

**Reply by favourwright** [reply]
having the same issue highlighted and this didn't fix it

**Comment by wildplant** [comment]
I commented the extractCss option in `nuxt.config.js`,  the css invert works.
``` js
  // Build Configuration: https://go.nuxtjs.dev/config-build
  build: {
    // extractCSS: true,
  }
}
```

**Comment by tomhatzer** [comment]
Hey everyone!

I have the same problem using filters (filter and backdrop-filter). It works on my local machine in dev mode but it won't work when running it in production.
I'm using NuxtJS with tailwindcss 3.0.23 and the @nuxtjs/tailwindcss plugin.

What fixed it for me was adding a custom class where the filter is set by me like (not my preferred solution but it works):
```
.custom-filter {
    filter: blur(4px);
}
```

Somehow empty variables get stripped in the production build process of NuxtJS or a plugin that's used in there.

Thank you for reading.

Have a great day!

Best wishes
Tom

**Comment by ghost** [comment]
Hello. I don't use nuxt.js but I had a similar problem. 

Fix was to *not minify* CSS.

I use gulp scripts to run development and build steps.

For dev: I use a `<link>` tag to include tailwind's output.css (so CSS straight from tailwind cli). There, backdrop works as expected.

For production: gulp first inlines the CSS between `<style>` tags then a gulp plugin called `gulp-htmlmin` minifies html, which had `minifyCSS: true` in its pipe options.

It turns out, minification breaks the backdrop blur for me and setting it to `minifyCSS: false` fixed the problem.

I hope this can help someone.

Good days!

**Comment by arinker** [comment]
Chrome (Version 98.0.4758.109) on Mac obviously has a bug.

The drop-shadows based on the filter utility do not work as expected. The css var defaults are ignored.
<img width="489" alt="Bildschirmfoto 2022-03-07 um 09 33 04" src="https://user-images.githubusercontent.com/2835005/156997540-161c1485-a888-4878-b026-132a5b2c4d37.png">

Once you check one css var in the developer tools all css var defaults are respected.
<img width="490" alt="Bildschirmfoto 2022-03-07 um 09 33 17" src="https://user-images.githubusercontent.com/2835005/156997572-7d619dce-7f12-4404-b698-29132f65d4d9.png">

**Comment by marijnbent** [comment]
I have the same issue. Css error because of the empty values for some vars.

CSS:
```
@tailwind base;

@tailwind components;

@tailwind utilities;

```
I don't even use filters, so I don't understand why these variables are added.

config:
```
module.exports = {
	content: ['./resources/**/*.{html,js}'],
	corePlugins: {
		preflight: false,
	},
	theme: {
		extend: {},
	},
};

```

The exact same setup, where I change content for purge with Tailwind 2 works correctly. The variables are added without causing an error in the output.

**Comment by Torsten85** [comment]
Also encountered this bug today. 
My nextjs app didn't show any backdrop filters after build because auf the missing `--tw-backdrop-sepia` definition.
It seems like somewhere (maybe in postcss, maybe in sass) during build, the **last(!)** empty variable declaration gets dropped.

I worked around it by changing my default style.sass to this:
```scss
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
  * {
    --tw-backdrop-sepia:  ;
    --will-be-deleted: ; 
  }
}
```

You'll see that the last definition `--will-be-deleted` won't land in the bundled output, but `--tw-backdrop-sepia` does.

**Comment by chrism** [comment]
I just encountered this issue in production too.

Using postcss (8.4.12) with cssnano (5.1.5) with latest (3.0.23) TailwindCSS.

Filters worked in development, not in production.

Fixed (for now) by downgrading TailwindCSS to 3.0.11 from advice in this thread
https://github.com/tailwindlabs/tailwindcss/discussions/7131

This definitely seems like a bug to me.

**Comment by rayasabari** [comment]
I have  the same issue, 
backdrop-blur work in developmnet, but not in production build,

I tried downgrading to tailwind 3.0.11 but still not works,
I also tried to downgrade the postcss to 8.4.5 and autoprefixer to  10.4.1 (because my last project using this version and it works) but still not fix my new projects.

can anyone tell the proper way to downgrade it? maybe I missed some step.

**Reply by Torsten85** [reply]
have you tried my workaround? Works for me without downgrading anything

**Reply by rayasabari** [reply]
Hi Sir, thanks for reply,
I also tried your way but unfortunately it's doesn't works too,

and now i solved it by changing the target from 'static' to 'server' in nuxt.config.js

**Reply by rayasabari** [reply]
my current issue:

-  when I run generate and start in local with `ssr: true` and `target: 'server'` in nuxt.config.js, the backdrop works, but when I deploy to vercel it's doesn't works

- I tried `ssr: false `and set `target: 'server'` then the backdrop works in vercel, but I don't want it, I want ssr still enabled 

UPDTAE:
Solved, I forgot to adding a `vercel.json` file in root project

**Comment by sefrijn** [comment]
Same issue using Nuxt JS 2.15.8, Postcss 8.4.12 and Tailwind 3.0.23.
Problem doesn't go away with downgrading to TW 3.0.10 here.

`backdrop-blur-lg` is not working in nuxt static build

**Reply by dandn9** [reply]
I downgraded to 3.0.11 and it worked

**Comment by fshareef** [comment]
Downgrading to 3.0.10 fixed issues for me with --tw-xx filter variables being removed.

**Comment by adamwathan** [comment]
If this is an issue for anyone, the best way to actually push things forward is to open an issue that includes a minimal reproduction 👍 We aren't seeing this on our own projects, and any time we've looked into it the real issue has been a bug in whatever CSS minifier was being used. But we have no issues that we've seen using cssnano for our stuff.

**Reply by designchapnl** [reply]
Hi @adamwathan , thanks for your reply and for your awesome framework!

I opened this as an issue, but it got moved to discussions somehow.

As a check I created a new Nuxtjs project with only Tailwindcss included like the Tailwindcss doc explains: https://tailwindcss.com/docs/guides/nuxtjs

I created a div with some text and a div overlay with blur applied to it.
When running dev server all goes well.
<img width="887" alt="dev" src="https://user-images.githubusercontent.com/29886711/169826082-576a2f20-2e5d-4107-ab28-b1ee2b0ad74b.png">
<img width="528" alt="css-dev" src="https://user-images.githubusercontent.com/29886711/169826127-b33152c9-3cda-4aa6-b030-c19acc8e52c7.png">

But when generating static html/css it doesn't apply the blur anymore:
<img width="891" alt="build" src="https://user-images.githubusercontent.com/29886711/169826254-b7747f42-1713-4ee6-aab7-5f46732fb693.png">

The only difference I see is this:
<img width="541" alt="css-build" src="https://user-images.githubusercontent.com/29886711/169826314-324ad585-aa2c-4d73-a6b4-ee797e0a17f5.png">

Could that be the issue? Adding or removing `backdrop-filter` does nothing.

Hope you guys can use this input. My enviroment is as statet in my first comment.

**Comment by sintj** [comment]
As on tailwind 3.0.24 the issue is still present!

**Reply by alpayoguz** [reply]
Same here!

**Reply by tobyab** [reply]
Yup! Still running into it too

**Reply by michael-andreuzza** [reply]
here too!

**Comment by startupfoundry** [comment]
I had this issue while using Vite. I upgraded to version "2.9.9" and the issue resolved itself.

**Comment by nozloy** [comment]
"tailwindcss": "^3.1.2"
"react": "^18.1.0"
"react-scripts": "5.0.1"

Some problem. Backdrop-blur doesnt work on production build

**Comment by Akryum** [comment]
Issue still present in Tailwind 3.1 and Nuxt 2
I manged to "fix" it by extracting the CSS and disabling cssnano:

![image](https://user-images.githubusercontent.com/2798204/173264690-061c9873-8ba3-4efe-9803-88596ca94959.png)

**Comment by jasperverbeet** [comment]
**TL;DR**: add `"cssnano": "^5.1.12"` to your peer dependencies in your `package.json`.

Issue prevails, but this doesn't seem to be a Tailwind issue but a Nuxt issue (more specifically a `cssnano` issue) as some have already mentioned. What happens is that some unused CSS variables are stripped out by `cssnano` during a Nuxt production build, causing styles that depend on them to disappear. I figured this out from this issue: https://github.com/tailwindlabs/tailwindcss/issues/7802.

- Turning off `optimizeCSS` is **not** the way to go, as that will obviously result in larger bundle sizes. (https://github.com/tailwindlabs/tailwindcss/discussions/7044#discussioncomment-2934845)
- Downgrading TailwindCSS might work but is not a complete solution for me. (https://github.com/tailwindlabs/tailwindcss/discussions/7044#discussioncomment-2468810)
- Defining the missing variables in a custom .css file did not do the trick for me as those were also stripped out by cssnano during a production build. (https://github.com/tailwindlabs/tailwindcss/discussions/7044#discussioncomment-2172498)

Running `npm explain cssnano` provides some insight on where it's used: `optimize-css-assets-webpack-plugin`, which is probably installed in the default `@nuxt/webpack` configuration.

```bash
> npm explain cssnano

cssnano@4.1.11
node_modules/optimize-css-assets-webpack-plugin/node_modules/cssnano
  cssnano@"^4.1.10" from optimize-css-assets-webpack-plugin@5.0.8
  node_modules/optimize-css-assets-webpack-plugin
    optimize-css-assets-webpack-plugin@"^5.0.4" from @nuxt/webpack@2.15.8
    node_modules/@nuxt/webpack
      @nuxt/webpack@"2.15.8" from @nuxt/builder@2.15.8
      node_modules/@nuxt/builder
        @nuxt/builder@"2.15.8" from nuxt@2.15.8
        node_modules/nuxt
          nuxt@"^2.15.8" from the root project
          peer nuxt@"^2.15" from @nuxtjs/composition-api@0.25.2
          node_modules/@nuxtjs/composition-api
            dev @nuxtjs/composition-api@"^0.25.2" from the root project
      @nuxt/webpack@"2.15.8" from nuxt@2.15.8
      node_modules/nuxt
        nuxt@"^2.15.8" from the root project
        peer nuxt@"^2.15" from @nuxtjs/composition-api@0.25.2
        node_modules/@nuxtjs/composition-api
          dev @nuxtjs/composition-api@"^0.25.2" from the root project
```

There recently has been in a fix in `cssnano@5.1.2` (https://github.com/cssnano/cssnano/commit/4ed396795a88d8b06953b9abb2656f686402821c): `fix(postcss-discard-empty): preserve empty custom properties`, which will now actually preserve empty variables and thus fix the issues mentioned in this thread. Upgrading `cssnano` to version 5.1.2 resolved everything for me.

The best way to do this is by upgrading the peer dependencies of [optimize-css-assets-webpack-plugin](https://www.npmjs.com/package/optimize-css-assets-webpack-plugin), but that package doesn't seem to be updated anymore. In the meantime you can simply upgrade the peer dependencies in your own `package.json`:

```json
"peerDependencies": {
    "cssnano": "^5.1.12"
}
```

If the issue is still there you might want to add the following variables to your main.css/tailwind.css file to make sure they're actually present:

```
*,
::before,
::after {
  --tw-backdrop-blur: ;
  --tw-backdrop-brightness: ;
  --tw-backdrop-contrast: ;
  --tw-backdrop-grayscale: ;
  --tw-backdrop-hue-rotate: ;
  --tw-backdrop-invert: ;
  --tw-backdrop-opacity: ;
  --tw-backdrop-saturate: ;
  --tw-backdrop-sepia: ;
}
```

**Reply by nicolas-giuristante** [reply]
Thank you @jasperverbeet, this was insightful and I thought it would actually fix the issue. Although I was not able to fix it by simply adding a peerDependencies.

My NuxtJS is pretty minimal I'm not sure what could be interfering. Here is what it looks like after adding the dependency and running npm explain cssnano.
```
cssnano@4.1.11
node_modules/@nuxt/webpack/node_modules/cssnano
  cssnano@"^4.1.11" from @nuxt/webpack@2.15.8
  node_modules/@nuxt/webpack
    @nuxt/webpack@"2.15.8" from @nuxt/builder@2.15.8
    node_modules/@nuxt/builder
      @nuxt/builder@"2.15.8" from nuxt@2.15.8
      node_modules/nuxt
        nuxt@"^2.15.8" from the root project
    @nuxt/webpack@"2.15.8" from nuxt@2.15.8
    node_modules/nuxt
      nuxt@"^2.15.8" from the root project

cssnano@5.1.12 peer
node_modules/cssnano
  peer cssnano@"^5.1.12" from the root project

cssnano@4.1.11
node_modules/optimize-css-assets-webpack-plugin/node_modules/cssnano
  cssnano@"^4.1.10" from optimize-css-assets-webpack-plugin@5.0.8
  node_modules/optimize-css-assets-webpack-plugin
    optimize-css-assets-webpack-plugin@"^5.0.4" from @nuxt/webpack@2.15.8
    node_modules/@nuxt/webpack
      @nuxt/webpack@"2.15.8" from @nuxt/builder@2.15.8
      node_modules/@nuxt/builder
        @nuxt/builder@"2.15.8" from nuxt@2.15.8
        node_modules/nuxt
          nuxt@"^2.15.8" from the root project
      @nuxt/webpack@"2.15.8" from nuxt@2.15.8
      node_modules/nuxt
        nuxt@"^2.15.8" from the root project
```

I am using the same nuxt version as you (2.15.8). I am curious what package and version of TailwindCSS you're using. I am using @nuxtjs/tailwindcss (5.1.3) Are you using tailwindlabs/tailwindcss? Do you think it matters what package I am using?

Cheers!

**Reply by jasperverbeet** [reply]
Hmmm weird, maybe I mixed things up or something. I created a local prototype and the issue disappeared for me after I added the peerDependency. 

The issue was introduced in 3.0.12: https://github.com/tailwindlabs/tailwindcss/compare/v3.0.11...v3.0.12, somehow the empty CSS variables with the Nuxt CSS optimizer remove the empty variables entirely.

**Reply by tomhatzer** [reply]
It may possibly have to do with refactoring defaults in the commit https://github.com/tailwindlabs/tailwindcss/commit/f2d73b8c3d6f028680bd0b6239102be231b4ff46 from https://github.com/tailwindlabs/tailwindcss/pull/6926 ?

**Comment by Kayrim** [comment]
Same issue, tried a few different solutions on this thread but in the end the only that worked was downgrading to 3.0.10 hopefully this gets picked up soon but for now it will have to do.

**Reply by coaperator** [reply]
downgrading to version 3.0.11 helped me

**Comment by cchacholiades** [comment]
Followed @Akryum's suggestion, these options did it for me:

```
// nuxt.config.js

minifyCSS: true,
minifyJS: true,
extractCSS: true,
optimizeCSS: false
```

NuxtJS + tailwind

```
"nuxt": "^2.15.7",
"@nuxtjs/tailwindcss": "^5.3.0",
```

**Comment by PaulHarrer** [comment]
Same issue here but even with minimum setup like in tailwind documentation (with tailwind cli). Only downgrading tailwind version to 3.0.10 - as mentioned by others - is working.

**Reply by davidhellmann** [reply]
Downgrade also not working here within a nuxt 2. In my case I use dropshadow and grayscale filters...

**Comment by mcenatie** [comment]
Same issue here with create-react-app. When building our app, all the empty custom variables like --tw-backdrop-brightness, --tw-backdrop-contrast got deleted from the css, which was causing issues with the backdrop filter. The workaround we ended up using, was to add a custom plugin in the tailwind.config.js to override the existing backdrop-blur plugin from tailwind. Basically we just reused the one from tailwind but updated it by adding a fallback for each variable (--tw-variable-name,) 

```
plugin(function({ matchUtilities, theme }) {
            matchUtilities(
                {
                    'backdrop-blur': value => {
                        const cssBackdropFilterValue = [
                            'var(--tw-backdrop-blur,)',
                            'var(--tw-backdrop-brightness,)',
                            'var(--tw-backdrop-contrast,)',
                            'var(--tw-backdrop-grayscale,)',
                            'var(--tw-backdrop-hue-rotate,)',
                            'var(--tw-backdrop-invert,)',
                            'var(--tw-backdrop-opacity,)',
                            'var(--tw-backdrop-saturate,)',
                            'var(--tw-backdrop-sepia,)',
                        ].join(' ');

                        return {
                            '--tw-backdrop-blur': `blur(${value})`,
                            '@defaults backdrop-filter': {},
                            'backdrop-filter': cssBackdropFilterValue,
                        };
                    },
                },
                { values: theme('backdropBlur') }
            );
        }),
```

In that way, even if the variable is not defined in the css, there is a fallback that is set to an empty value. More details about this trick here: https://css-tricks.com/optional-custom-property-values-trick/
I guess the tailwind team could apply this fix in their own plugin...

**Reply by iyadthayyil** [reply]
Thanks, @mcenatie, nice workaround.
It solves the issue without downgrading!

**Reply by sbdchd** [reply]
With this I run into:

```
warn - The class `after:backdrop-blur-[6px]` is ambiguous and matches multiple utilities.
warn - If this is content and not a class, replace it with `after:backdrop-blur-&lsqb;6px&rsqb;` to silence this warning.
```

workaround is to update it to a unique name, `like backdrop2-blur`

**Comment by HiOliver** [comment]
Ended up using the following solution so the minifier wouldn't consider : ; as invalid which finally worked for a production build:

```
:root {
    --empty: ;
}
*,
:after,
:before {
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-pan-x: var(--empty, /*!*/ /*!*/);
    --tw-pan-y: var(--empty, /*!*/ /*!*/);
    --tw-pinch-zoom: var(--empty, /*!*/ /*!*/);
    --tw-scroll-snap-strictness: proximity;
    --tw-ordinal: var(--empty, /*!*/ /*!*/);
    --tw-slashed-zero: var(--empty, /*!*/ /*!*/);
    --tw-numeric-figure: var(--empty, /*!*/ /*!*/);
    --tw-numeric-spacing: var(--empty, /*!*/ /*!*/);
    --tw-numeric-fraction: var(--empty, /*!*/ /*!*/);
    --tw-ring-inset: var(--empty, /*!*/ /*!*/);
    --tw-ring-offset-width: 0px;
    --tw-ring-offset-color: #fff;
    --tw-ring-color: rgba(59, 130, 246, 0.5);
    --tw-ring-offset-shadow: 0 0 #0000;
    --tw-ring-shadow: 0 0 #0000;
    --tw-shadow: 0 0 #0000;
    --tw-shadow-colored: 0 0 #0000;
    --tw-blur: var(--empty, /*!*/ /*!*/);
    --tw-brightness: var(--empty, /*!*/ /*!*/);
    --tw-contrast: var(--empty, /*!*/ /*!*/);
    --tw-grayscale: var(--empty, /*!*/ /*!*/);
    --tw-hue-rotate: var(--empty, /*!*/ /*!*/);
    --tw-invert: var(--empty, /*!*/ /*!*/);
    --tw-saturate: var(--empty, /*!*/ /*!*/);
    --tw-sepia: var(--empty, /*!*/ /*!*/);
    --tw-drop-shadow: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-blur: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-brightness: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-contrast: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-grayscale: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-hue-rotate: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-invert: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-opacity: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-saturate: var(--empty, /*!*/ /*!*/);
    --tw-backdrop-sepia: var(--empty, /*!*/ /*!*/);
}

```

Taken from: https://github.com/tailwindlabs/tailwindcss/issues/7121#issuecomment-1026823461

**Reply by shanecarmody** [reply]
Worked for me, thanks!

**Comment by token-ed** [comment]
Hey guys, hi @adamwathan. First of all, thank you for this amazing framework, this is the kind of stuff that changes the world.

It  seems I can't make `backdrop-blur-sm` to work in my project (using NextJS)

I've pushed a minimal reproducible version of a brand new NextJS project using `create-next-app` (to be sure the configurations are immaculate) and it seems it's not working there as well - I've pushed a repo to GitHub and you can find it here: https://github.com/token-ed/test

Also deployed it with GitHub pages for immediate access if you want to take a look at those generated classes: [Test Deploy](https://token-ed.github.io/test/)

Here's a sample screenshot:
![image](https://github.com/tailwindlabs/tailwindcss/assets/149210610/212f50db-28ed-4439-8d9c-b358e4fffa96)

Only thing needed to run the project locally is `npm run dev` and it should be all set.

All help would be much appreciated :pray: Cheers.

### EDIT :writing_hand: 
Looks like I rushed it. `backdrop-blur-[...]` yields to the CSS `backdrop-filter`. Obviously this property is used to blur whatever is behind a certain element. In my example there is no other elements other than the one rendered - that's why we don't see it blurred. Either I would have to use `blur` or have a parent element covering the child with a `backdrop-blur-[...]` set on it.
So this was a false alarm. Leaving this here for reference.
